lib/mayaUsd/fileio/primUpdaterManager.cpp

```diff
@@ -1724,4 +1747,270 @@ bool PrimUpdaterManager::readPullInformation(const MDagPath& dagPath, Ufe::Path&
+{
+    TF_VERIFY(_pulledPrims.root()->empty());
+    _orphanedNodesManager = std::make_shared<OrphanedNodesManager>(_pulledPrims);
+    Ufe::Scene::instance().addObserver(_orphanedNodesManager);
+
+    // Observe Maya so we can stop scene observation on file new or open.
+    MStatus                status;
+    MSceneMessage::Message msgs[] = { MSceneMessage::kBeforeNew, MSceneMessage::kBeforeOpen };
+    for (auto msg : msgs) {
+        _fileCbs.append(MSceneMessage::addCallback(msg, beforeNewOrOpenCallback, this, &status));
+        CHECK_MSTATUS(status);
+    }
+}
+
+void PrimUpdaterManager::endManagePulledPrims()
+{
+    TF_VERIFY(_orphanedNodesManager);
+    TF_VERIFY(Ufe::Scene::instance().removeObserver(_orphanedNodesManager));
+    auto status = MMessage::removeCallbacks(_fileCbs);
+    CHECK_MSTATUS(status);
+    _fileCbs.clear();
+    _orphanedNodesManager.reset();
+    _pulledPrims.clear();
+}
+
+/*static*/
+void PrimUpdaterManager::beforeNewOrOpenCallback(void* clientData)
+{
+    auto* pum = static_cast<PrimUpdaterManager*>(clientData);
+    pum->endManagePulledPrims();
+}
+
+//==============================================================================
+//
+// CLASS PrimUpdaterManager::OrphanedNodesManager
+//
+//==============================================================================
+
+PrimUpdaterManager::OrphanedNodesManager::OrphanedNodesManager(
+    const Ufe::Trie<PullVariantInfo>& pulledPrims)
+    : _pulledPrims(pulledPrims)
+{
+}
+
+void PrimUpdaterManager::OrphanedNodesManager::operator()(const Ufe::Notification& n)
+{
+    const auto& sceneNotification = static_cast<const Ufe::SceneChanged&>(n);
+    auto        changedPath = sceneNotification.changedPath();
+
+    // No changedPath means composite.  Use containsDescendant(), as
+    // containsDescendantInclusive() would mean a structure change on the
+    // pulled node itself, which is not possible (pulled objects are locked).
+    if (changedPath.empty()) {
+        const auto& sceneCompositeNotification
+            = static_cast<const Ufe::SceneCompositeNotification&>(n);
+        for (const auto& op : sceneCompositeNotification.opsList()) {
+            if (_pulledPrims.containsDescendant(op.path)) {
+                handleOp(op);
+            }
+        }
+    } else if (_pulledPrims.containsDescendant(changedPath)) {
+        handleOp(sceneNotification);
+    }
+}
+
+void PrimUpdaterManager::OrphanedNodesManager::handleOp(
+    const Ufe::SceneCompositeNotification::Op& op)
+{
+    switch (op.opType) {
+    case Ufe::SceneChanged::ObjectAdd: {
+        // Restoring a previously-deleted scene item may restore an orphaned
+        // node.  Traverse the trie, and show hidden pull parents that are
+        // descendants of the argument path.  The trie node that corresponds to
+        // the added path is the starting point.  It may be an internal node,
+        // without data.
+        auto ancestorNode = _pulledPrims.node(op.path);
```

Comment:
I"m not going to try to figure it out now, but a given prim could be hidden both because one of its variant change *and* it got deleted afterwards. So I think restoring visibility (here and for variant switch below in case multiple variants were switched...) need to check that all original causes why it was hidden are now gone.